src/puppetlabs/puppetdb/cli/shovel.clj

```diff
@@ -0,0 +1,375 @@
+      (catch java.io.EOFException e
+        (log/warn (i18n/trs "Caught EOFException on broker startup, trying again. This is probably due to KahaDB corruption (see \"KahaDB Corruption\" in the PuppetDB manual)."))
+        (retry-build-and-start-broker! brokername dir config)))
+    (catch java.io.IOException e
+      (throw (java.io.IOException.
+              (i18n/trs "Unable to start broker in {0}. This is probably due to KahaDB corruption or version incompatibility after a PuppetDB downgrade (see \"KahaDB Corruption\" in the PuppetDB manual)." (pr-str dir))
+              e)))))
+
+(defn activemq-connection-factory [spec]
+  (proxy [PooledConnectionFactory java.lang.AutoCloseable]
+      [(ActiveMQConnectionFactory. spec)]
+    (close [] (.stop this))))
+
+(defn upgrade-lockfile-path [vardir]
+  (io/file vardir "mq-migrated"))
+
+(defn needs-upgrade?
+  "Returns true if the user has an unmigrated ActiveMQ director in
+  their vardir"
+  [config]
+  (let [vardir (get-in config [:global :vardir])]
+    (and (not (fs/exists? (upgrade-lockfile-path vardir)))
+         (fs/exists? (io/file vardir "mq")))))
+
+(defn lock-upgrade
+  "Puts a lockfile in vardir to indicate the user has already migrated
+  to Stockpile"
+  [config]
+  (fs/touch (upgrade-lockfile-path (get-in config [:global :vardir]))))
+
+;;
+
+(defn activemq->stockpile
+  "Drains the queue found in `cmd-proc-config` and uses `enqueue-fn`
+  to resubmit those commands to PuppetDB (running Stockpile). All
+  failures to enqueue are discarded using the `dlo`"
+  [{global-config :global
+    {mq-config :mq :as cmd-proc-config} :command-processing}
+   enqueue-fn
+   dlo]
+  ;; Starting
+  (let [mq-dir (str (io/file (:vardir global-config) "mq"))
+        mq-broker-url (format "%s&wireFormat.maxFrameSize=%s&marshal=true"
+                              (:address mq-config "vm://localhost?jms.prefetchPolicy.all=1&create=false")
+                              (:max-frame-size cmd-proc-config 209715200))
+        mq-discard-dir (str (io/file mq-dir "discard"))
+        mq-endpoint (:endpoint mq-config default-mq-endpoint)
+        conn-pool (activemq-connection-factory mq-broker-url)
+        broker (do (log/info (i18n/trs "Starting broker"))
+                   (build-and-start-broker! "localhost" mq-dir cmd-proc-config))
+        process-message (create-message-processor enqueue-fn dlo)
+        id (atom 0)]
+
+    (try
+      ;; Drain the scheduler so we don't get any extra messages on the Queue when
+      ;; we're processing
+      (create-scheduler-receiver conn-pool mq-endpoint process-message dlo id)
+      (create-mq-receiver conn-pool mq-endpoint process-message id)
+      (create-mq-receiver conn-pool retired-mq-endpoint process-message id)
+
+      (finally
+        ;; Stopping
+
+        (log/info (i18n/trs "Stopping ActiveMQConnectionFactory"))
+        (when conn-pool
+          (.stop conn-pool))
+
+        (log/info (i18n/trs "Stopping broker"))
+        (when broker
```

Comment:
Wondered if these whens were necessary, or maybe it's handling the case where the constructors fail, and the attempt to use them in the try detonates; in any case, wondered whether nesting finally blocks like this might could make sense (assumes we do need th whens):

    (let [mq-dir (str (io/file (:vardir global-config) "mq"))
          mq-broker-url ...
          mq-discard-dir ...
          mq-endpoint ...
          process-message ...
          id (atom 0)]
      (when-let [conn-pool (activemq-connection-factory mq-broker-url)]
        (try
          (when-let [broker (do ...)]
            (try
              (create-scheduler-receiver conn-pool mq-endpoint process-message dlo id)
              ...
              (finally
                (stop-broker! broker))))
          (finally
            (log/info (i18n/trs "Stopping ActiveMQConnectionFactory"))
            (.stop conn-pool)))))